server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -626,7 +627,7 @@ private SearchPhaseResult executeQueryPhase(ShardSearchRequest request, SearchSh
             } finally {
                 tracer.stopTrace();
             }
-            if (request.numberOfShards() == 1) {
+            if (request.numberOfShards() == 1 && (request.source() == null || request.source().rankBuilder() == null)) {
```

Comment:
This avoids short cutting to fetch if were ranking because we still need to do the final ranking on the coordinator. I considered doing the ranking here and allowing the shortcut, but when I tried it, it was quite a bit more complexity.

Reply:
Got it, that makes sense. Would it make sense to have a method somewhere called "canShortcutFetch" or a comment that explains the additional conditionals?

Reply:
I think both are good ideas, and I will tackle this in a follow up if that's okay.